shuttle/src/future/mod.rs

```diff
@@ -11,44 +11,49 @@ use crate::runtime::thread;
 use std::error::Error;
 use std::fmt::{Display, Formatter};
 use std::future::Future;
+use std::panic::Location;
 use std::pin::Pin;
 use std::result::Result;
 use std::sync::Arc;
 use std::task::{Context, Poll, Waker};
 
 pub mod batch_semaphore;
 
-fn spawn_inner<F>(fut: F) -> JoinHandle<F::Output>
+fn spawn_inner<F>(fut: F, caller: &'static Location<'static>) -> JoinHandle<F::Output>
 where
     F: Future + 'static,
     F::Output: 'static,
 {
     let stack_size = ExecutionState::with(|s| s.config.stack_size);
     let inner = Arc::new(std::sync::Mutex::new(JoinHandleInner::default()));
-    let task_id = ExecutionState::spawn_future(Wrapper::new(fut, inner.clone()), stack_size, None);
+    let task_id = ExecutionState::spawn_future(Wrapper::new(fut, inner.clone()), stack_size, None, caller);
 
     thread::switch();
 
     JoinHandle { task_id, inner }
 }
 
 /// Spawn a new async task that the executor will run to completion.
+#[track_caller]
 pub fn spawn<F>(fut: F) -> JoinHandle<F::Output>
 where
     F: Future + Send + 'static,
     F::Output: Send + 'static,
 {
-    spawn_inner(fut)
+    let caller = Location::caller();
+    spawn_inner(fut, caller)
 }
 
 /// Spawn a new async task that the executor will run to completion.
 /// This is just `spawn` without the `Send` bound, and it mirrors `spawn_local` from Tokio.
+#[track_caller]
 pub fn spawn_local<F>(fut: F) -> JoinHandle<F::Output>
 where
     F: Future + 'static,
     F::Output: 'static,
 {
-    spawn_inner(fut)
+    let caller = Location::caller();
+    spawn_inner(fut, caller)
```

Comment:
Nit on these two is to compact to `spawn_inner(fut, Location::caller())` (yes, very nitty, but it helps when reading the code to see quickly that the function is just a thin wrapper)